src/adapter/http/src/middleware/current_account_resolver_layer.rs

```diff
@@ -0,0 +1,136 @@
+                .unwrap();
+
+            match authentication_service
+                .account_info_by_token(access_token.token)
+                .await
+            {
+                Ok(account_info) => {
+                    Ok(CurrentAccountSubject::new(account_info.account_name, false))
+                }
+                Err(auth::GetAccountInfoError::AccessToken(_)) => Ok(CurrentAccountSubject::new(
+                    opendatafabric::AccountName::new_unchecked(auth::ANONYMOUS_ACCOUNT_NAME),
+                    true,
+                )),
+                Err(auth::GetAccountInfoError::Internal(_)) => {
+                    return Err(Response::builder()
+                        .status(StatusCode::INTERNAL_SERVER_ERROR)
+                        .body(Default::default())
+                        .unwrap());
+                }
+            }
+        } else {
+            Ok(CurrentAccountSubject::new(
+                opendatafabric::AccountName::new_unchecked(auth::ANONYMOUS_ACCOUNT_NAME),
+                true,
+            ))
+        }
+    }
+}
+
+impl<Svc> Service<Request<Body>> for CurrentAccountResolverMiddleware<Svc>
+where
+    Svc: Service<Request<Body>, Response = Response> + Send + 'static + Clone,
+    Svc::Future: Send + 'static,
+{
+    type Response = Svc::Response;
+    type Error = Svc::Error;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut request: Request<Body>) -> Self::Future {
+        // Inspired by https://github.com/maxcountryman/axum-login/blob/main/axum-login/src/auth.rs
+        // TODO: PERF: Is cloning a performance concern?
+        let mut inner = self.inner.clone();
+
+        Box::pin(async move {
+            let maybe_access_token = request
+                .extract_parts::<Option<
+                    axum::TypedHeader<
+                        axum::headers::Authorization<axum::headers::authorization::Bearer>,
+                    >,
+                >>()
+                .await
+                .unwrap()
+                .map(|th| AccessToken::new(th.token()));
+
+            let base_catalog = request
+                .extensions()
+                .get::<dill::Catalog>()
+                .expect("Catalog not found in http server extensions");
+
+            let current_account_subject =
+                match Self::current_account_subject(base_catalog, maybe_access_token).await {
+                    Ok(current_account_subject) => current_account_subject,
+                    Err(response) => return Ok(response),
+                };
+
+            let derived_catalog = dill::CatalogBuilder::new_chained(base_catalog)
+                .add_value(current_account_subject)
+                .build();
+
+            request.extensions_mut().insert(derived_catalog.clone());
```

Comment:
Why do we need to clone here? Doesn't look like original instance is needed any more.

Reply:
Eliminated